oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/jwt/JwtVerifier.java

```diff
@@ -1,3 +1,18 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 package org.springframework.security.oauth2.jwt;
 
 /**
```

Comment:
This operation `void verifyClaims(Jwt jwt)` only works for `Jwt` Access Tokens. We need to ensure it also works for Opaque tokens as we'll need to provide support for [Token Introspection](https://tools.ietf.org/search/rfc7662).

Reply:
The extension point I'm thinking of is actually in a different spot, which is the yet-to-be-created `AbstractOAuth2AccessTokenAuthenticationProvider` which `JwtEncodedOAuth2AccessTokenAuthenticationProvider` extends from.

It has a `mapClaims` method which I would imagine does something (possibly) much different than the JwtEncoded implementation.

I haven't spent a lot of time fleshing that out, though. You have probably seen further down the road than I.

Reply:
We are going to need a facility for verifying Access Token attributes in general.

For example, for Opaque Access Tokens, where we implement Token Introspection, a user may want to verify one or more of those attributes returned from the Token Introspection endpoint. 

So something like this would work for both Jwt and Opaque tokens:

```
public interface OAuth2TokenVerifier {
	boolean verify(Map<String, Object> tokenAttributes);
}
```

This could also work if we needed to provide a hook for user's wanting to validate claims in the ID Token as well.